Default es metadata javascript transforms

AndreKurait opened the pull request:
### Description

Adds a framework for version specific javascript transforms in metadata migrations.
Adds logic for version specific transforms along with inclusion in result output and a transforms file log which contains the transform json, 

User Experience:

```
...
Transformations:
   dense_vector to knn_vector:
      Convert mapping type dense_vector to OpenSearch knn_vector
   Version Transform:
      Other transforms for source to target conversion

Results:
   0 issue(s) detected

Consult /shared-logs-output/migration-console-default/ip-10-0-166-104.ec2.internal/metadata/metadata2025-07-28_18-12-09.log to see detailed logs for this run
See transformations applied for this run at /shared-logs-output/migration-console-default/ip-10-0-166-104.ec2.internal/metadata/metadata2025-07-28_18-12-09-transforms.json
```

And contents of the file:
```
(19:45:40) migration-console (~) -> cat /shared-logs-output/migration-console-default/ip-10-0-172-2.ec2.internal/metadata/metadata2025-07-28_19-45-34-transforms.log
Default breaking changes transform config:
[ {
  "JsonJSTransformerProvider" : {
    "initializationResourcePath" : "js/es-vector-knn-metadata.js",
    "bindingsObject" : "{}"
  }
} ]
```

When custom transforms are specified:
```
Transformations:
   User Supplied Custom Transform:
      Custom transformation applied from supplied arguments.
      Skipped default version-specific transformations:
      (Field Data Type Deprecation - string)
   Version Transform:
      Other transforms for source to target conversion
```

### Issues Resolved
<!-- List any GitHub or Jira issues this PR will resolve -->

### Testing
Unit tests added, verified logging to EFS in MA

### Check List
- [x] New functionality includes testing
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.
    TODO, with links in metadata output, will include this in a followup PR

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

peternied reviewed (commented):
1/ How would would a customer see the transformation if they wanted to edit it? Seem like there isn't a way without pulling up our codebase.

2/ The additional log file is useful, but I think we should have something human readable that appears during metadata migration.

```
Transformation:
    Convert deprecated field type string to text - the source cluster is running version Elasticsearch 5 or lower and a field with the data type 'string' was found.
        Learn more https://github.com/opensearch-project/opensearch-migrations/blob/main/docs/breaking-changes/StringMappingType.md
    Translating vector field type to knn - the source cluster is running Elasticsearch 7 or greater and a field with the data type 'dense_vector' was found.
        Learn more https://docs.opensearch.org/latest/migration-assistant/migration-phases/planning-your-migration/handling-field-type-breaking-changes/
...
```

3/ How would a customer know that the default transformations aren't going to be applied if they supplied their own file?  I think we need something that clearly bridges this gap too.

AndreKurait commented:
> 1/ How would would a customer see the transformation if they wanted to edit it? Seem like there isn't a way without pulling up our codebase.
> 
> 2/ The additional log file is useful, but I think we should have something human readable that appears during metadata migration.
> 
> ```
> Transformation:
>     Convert deprecated field type string to text - the source cluster is running version Elasticsearch 5 or lower and a field with the data type 'string' was found.
>         Learn more https://github.com/opensearch-project/opensearch-migrations/blob/main/docs/breaking-changes/StringMappingType.md
>     Translating vector field type to knn - the source cluster is running Elasticsearch 7 or greater and a field with the data type 'dense_vector' was found.
>         Learn more https://docs.opensearch.org/latest/migration-assistant/migration-phases/planning-your-migration/handling-field-type-breaking-changes/
> ...
> ```
> 
> 3/ How would a customer know that the default transformations aren't going to be applied if they supplied their own file? I think we need something that clearly bridges this gap too.

1. We have info and url on the transformation so if we have one where user can supply config to, then they will be able to supply that (and we can build syntactic sugar in the form of the providers). I see a customer modifying the JS of a built in transformation as an advanced developer user and similar to one that is modifying arbitrary parts of the rest of our codebase.
2. Addressed this, see overview
3. Added logging in transformation output for this, see overview

AndreKurait commented on MetadataMigration/src/main/java/org/opensearch/migrations/commands/MigratorEvaluatorBase.java:
```diff
@@ -55,37 +70,111 @@ protected Clusters createClusters() {
+        return versionSpecificCustomTransforms;
+    }
+
+    protected void logTransformerConfig(String title, String transformerConfig) {
```
I don't necessarily want a transformation file for every single RFS process since there may be a ton over a run, This is something we'll get better support for with argo workflows

AndreKurait commented on MetadataMigration/src/main/java/org/opensearch/migrations/commands/MigratorEvaluatorBase.java:
```diff
@@ -55,37 +70,111 @@ protected Clusters createClusters() {
+            jsTransformationFiles.add(DENSE_VECTOR_KNN_TRANSFORMATION_FILE);
+            transformersBuilder.transformerInfo(Transformers.TransformerInfo
+                    .builder()
+                    .name("dense_vector to knn_vector")
```
This description and name is specific to metadata. I'd like to keep this in here until we added some friendly logging with these transformations to any of the other projects.

peternied commented on MetadataMigration/src/test/java/org/opensearch/migrations/ES8VectorFieldMappingsTransformationTest.java:
```diff
@@ -91,45 +88,25 @@ private void performCustomTransformationTest() {
         log.info(result.asCliOutput());
         assertThat(result.getExitCode(), equalTo(0));
 
+        assertThat(result.getTransformations().getTransformerInfos().size(), equalTo(2));
```
🔥

AndreKurait replied:
2nd being
```
   Version Transform:
      Other transforms for source to target conversion
```

peternied commented on transformation/standardJavascriptTransforms/src/es-flattened-flat-object.js:
Did you mean to include this file?
